src/main/java/org/apache/sysds/runtime/instructions/fed/TernaryFEDInstruction.java

```diff
@@ -0,0 +1,160 @@
+
+		if(matrixInputsCount == 3)
+			processMatrixInput(ec, mo1, mo2, mo3);
+		else if (matrixInputsCount == 1) {
+			CPOperand in = mo1 == null ? mo2 == null ? input3 : input2 : input1;
+			mo1 = mo1 == null ? mo2 == null ? mo3 : mo2 : mo1;
+			processMatrixScalarInput(ec, mo1, in);
+		} else
+			process2MatrixScalarInput(ec, mo1, mo2, mo3);
+	}
+
+	private void processMatrixScalarInput(ExecutionContext ec, MatrixObject mo1, CPOperand in) {
+		FederatedRequest fr1 = FederationUtils.callInstruction(instString, output,
+			new CPOperand[] {in}, new long[] {mo1.getFedMapping().getID()});
+		mo1.getFedMapping().execute(getTID(), true, fr1);
+
+		//derive new fed mapping for output
+		MatrixObject out = ec.getMatrixObject(output);
+		out.getDataCharacteristics().set(mo1.getDataCharacteristics());
+		out.setFedMapping(mo1.getFedMapping().copyWithNewID(fr1.getID()));
+	}
+
+	private void process2MatrixScalarInput(ExecutionContext ec, MatrixObject mo1, MatrixObject mo2, MatrixObject mo3) {
+		CPOperand[] inputArgs = new CPOperand[] {input1, input2};
+		if(mo1 != null && mo1.isFederated() && mo2 == null) {
+			mo2 = mo3;
+			inputArgs = new CPOperand[] {input1, input3};
+		} else if(mo2 != null && mo2.isFederated() && mo1 == null) {
+			mo1 = mo2;
+			mo2 = mo3;
+			inputArgs = new CPOperand[] {input2, input3};
+		} else if(mo2 != null && mo2.isFederated() && mo1 != null) {
+			mo1 = mo2;
+			mo2 = ec.getMatrixObject(input1);
+			inputArgs = new CPOperand[] {input2, input1};
+		} else if(mo3 != null && mo3.isFederated() && mo1 == null) {
+			mo1 = mo3;
+			inputArgs = new CPOperand[] {input3, input2};
+		} else if(mo3 != null && mo3.isFederated() && mo1 != null) {
+			mo1 = mo3;
+			mo2 = ec.getMatrixObject(input1);
+
+			inputArgs = new CPOperand[] {input3, input1};
+		}
+
+		FederatedRequest[] fr1 = mo1.getFedMapping().broadcastSliced(mo2, false);
+
+		FederatedRequest fr2 = FederationUtils.callInstruction(instString, output,
+			inputArgs, new long[] {mo1.getFedMapping().getID(), fr1[0].getID()});
+
+		FederatedRequest fr3 = mo1.getFedMapping().cleanup(getTID(), fr1[0].getID());
+		mo1.getFedMapping().execute(getTID(), true, fr1, fr2, fr3);
+
+		//derive new fed mapping for output
+		MatrixObject out = ec.getMatrixObject(output);
+		out.getDataCharacteristics().set(mo1.getDataCharacteristics());
+		out.setFedMapping(mo1.getFedMapping().copyWithNewID(fr3.getID()));
+	}
+
+
+	private void processMatrixInput(ExecutionContext ec, MatrixObject mo1, MatrixObject mo2, MatrixObject mo3) {
+		if(!mo1.isFederated())
+			if(mo2.isFederated()) {
+				mo1 = mo2;
+				mo2 = ec.getMatrixObject(input1);
+			} else {
+				mo1 = mo3;
+				mo3 = ec.getMatrixObject(input1);
+			}
+
+		FederatedRequest fr3;
+		// all 3 inputs aligned on the one worker
+		if(mo1.isFederated() && mo2.isFederated() && mo3.isFederated() && mo1.getFedMapping().isAligned(mo2.getFedMapping(), false) && mo1.getFedMapping().isAligned(mo3.getFedMapping(), false)) {
```

Comment:
> Thanks for the PR!
> It looks like a good start.
> 
> I think we could use some more tests that cover all branches of the processing of federated ternary instructions. One of the comments I added is regarding whether a part of the code can even be reached, so maybe we could think about a test that could cover this part (or if this is not possible, this part of the code could be removed).
> I think it would also be interesting to look at other ternary operations, for instance the "+_" and "-_". This is relevant for L2SVM and I have already looked at this in a separate branch with a solution that is targeted for this single purpose, but my approach is still incomplete, so it is more relevant to build this in your TernaryFEDInstruction version.

What do you mean by "+_" and "-_"? 
Heavy hitter instructions:
  #  Instruction  Time(s)  Count
  1  m_l2svm        0,137      1
  2  fed_ba+*       0,116     15
  3  fed_fedinit    0,061      1
  4  ba+*           0,022     29
  5  write          0,012      1
  6  rightIndex     0,011      1
  7  rmvar          0,003    459
  8  +*             0,002     35
  9  1-*            0,002     21
 10  createvar      0,002    227
 11  tak+*          0,002     28
 12  *              0,002     91
 13  list           0,001      4
 14  max            0,001     21
 15  tsmm           0,001     42
 16  r'             0,001     30
 17  >              0,001     14
 18  castdts        0,000     56
 19  -              0,000     35
 20  +              0,000     59

Reply:
I meant to write "+\*" and "-\*", but the stars were removed because it is a special character in this Github comment system :smile: . I can see that "+\*" is number 8 in your list of heavy hitter instructions.

Reply:
"+*" and "-*" are working, but the inputs in L2SVM are not federated. 
In while loop there is tmp_Xw = Xw + step_sz*Xd [line 115], Xd = X %*% s should be federated, but in AggregateBinaryFEDInstruction.java MV output is not federated [line 83]. I fixed this line with if( mo2.getNumColumns() == 1 && mo2.getNumRows() != mo1.getNumColumns()), but then out = 1 - Y * (tmp_Xw) [in line 84 inl2svm.dml] fails.

Reply:
Yes, this is also what I have been looking at. I am currently working on compiling federated instructions instead of converting them during runtime and how to control whether the output is federated. When doing this, I will look at the instructions in the inner loop of L2SVM and this is also the reason why I built a temporary TernaryFED and some additional cases for some of the other instructions. I am developing this in some other branches, so it is not necessary to have this in your PR. What I think is relevant for your PR is just to ensure that the "+\*" and "-\*" operations are supported in TernaryFEDInstruction. If this is implemented, I can later merge the compilation of FED instructions, including the federated output planning, with your version of TernaryFEDInstruction. If your changes break the L2SVM tests, you can decide not to convert the CP instruction to a FED instruction when it is a "+\*" or "-\*" operation.

Reply:
I fixed the L2SVM inner loop a few minutes ago, but I needed to modify a few fed instructions for this. So if you neeed, you can use it, but I do not know if other tests won't fail now 😄
But yes,  "+*" and "-*"  are implemented and supported.